crates/lib3h/src/gateway/gateway_transport.rs

```diff
@@ -72,50 +83,84 @@ impl P2pGateway {
+                        )))?;
                     }
-                    GhostCallbackData::Response(response) => response,
-                };
-                // Check if response is an error
-                let response = match response {
-                    Err(e) => {
-                        parent_msg.respond(Err(Lib3hError::new(ErrorKind::TransportError(e))))?;
-                        return Ok(());
+                    // No error but something other than SendMessageSuccess:
+                    GhostCallbackData::Response(Ok(_)) => {
+                        warn!(
+                            "Gateway got bad response type from transport: {:?}",
+                            response
+                        );
+                        cb(Err(format!("bad response type: {:?}", response).into()))?;
+                    }
+                    // Transport error:
+                    GhostCallbackData::Response(Err(_error)) => {
+                        debug!("Gateway got error from transport. Adding message to pending");
+                        me.pending_outgoing_messages.push(PendingOutgoingMessage {
+                            uri,
+                            payload,
+                            span: span.follower("pending due to error"),
+                            cb,
+                        });
+                    } //parent_msg.respond(Err(Lib3hError::new(ErrorKind::TransportError(e))))?,
+                    // Timeout:
+                    GhostCallbackData::Timeout => {
+                        debug!("Gateway got timeout from transport. Adding message to pending");
+                        me.pending_outgoing_messages.push(PendingOutgoingMessage {
+                            uri,
+                            payload,
+                            span: span.follower("pending due to timeout"),
+                            cb,
+                        });
                     }
-                    Ok(response) => response,
-                };
-                // Must be a SendMessage response
-                if let transport::protocol::RequestToChildResponse::SendMessageSuccess = response {
-                    parent_msg.respond(Ok(GatewayRequestToChildResponse::Transport(
-                        transport::protocol::RequestToChildResponse::SendMessageSuccess,
-                    )))?;
-                } else {
-                    parent_msg.respond(Err(format!("bad response type: {:?}", response).into()))?;
                 }
                 Ok(())
             }),
+            GhostTrackRequestOptions {
+                timeout: Some(std::time::Duration::from_millis(2000)),
+            },
         )
     }
 
+    pub(crate) fn handle_transport_pending_outgoing_messages(&mut self) -> GhostResult<()> {
+        let pending: Vec<PendingOutgoingMessage> =
+            self.pending_outgoing_messages.drain(..).collect();
```

Comment:
No need for `collect()` here.
```suggestion
            self.pending_outgoing_messages.drain(..);
```